File: Tuning/sample_code/15_io-format/src/fortran/main.f90
! Copyright 2024 Research Organization for Information Science and Technology
program main
  use mytype, only: SP, DP
  use timer, only: get_elp_time, get_cpu_time
  implicit none

  integer,parameter :: iunit = 11 
  integer,parameter :: NX=1024 !!2^10
  integer,parameter :: NY=1024 !!2^10
  integer :: NZ
  integer :: narg, ifmt
  integer :: i, j, k
  integer :: ielp0
  integer,allocatable,dimension(:,:,:) :: AA

  character(len=64) :: cbuf

  real(kind=DP) :: elp, elp0, cpu 
  real(kind=DP) :: aamemsize 

  !!write(6,*) SP, DP 

  narg = command_argument_count()
  if ( narg .ne. 2 ) then
     write(6,'("[usage] ./run.x (arg1) (arg2)",/,    &
&              "   (arg1): Integer (0-1)",/,         &
&              "           0: write w/  format",/,   &
&              "           1: write w/o format",/,   &
&              "   (arg2): Integer (<200) to change array size" &
&              )')
     stop
  end if

  call get_command_argument(1,cbuf)
  read(cbuf,'(1I10)') ifmt
  call get_command_argument(2,cbuf)
  read(cbuf,'(1I10)') NZ

  if ( NZ .gt. 200 ) then
     write(6,'(1A)') "[Error] 2nd arg must be < 200."
     stop 
  end if

  allocate( AA(NX,NY,NZ) ) 

  do k = 1, NZ
  do j = 1, NY
  do i = 1, NX
     AA(i,j,k) = i-1 + (j-1 + (k-1)*NY)*NX 
  end do
  end do
  end do

  aamemsize = 4.0_DP*NX*NY*NZ / 1048576 ! MiB

  !-------------------------------------------------------------------
  ! write with format
  !-------------------------------------------------------------------
  if ( ifmt .eq. 0 ) then

  cbuf="data1.out"
  open(iunit,FILE=cbuf) 
  call get_elp_time(ioffset=ielp0)  
  do k = 1, NZ
  do j = 1, NY
  do i = 1, NX
     write(iunit,'(1I10)') AA(i,j,k)
  end do
  end do
  end do
  call get_elp_time(ioffset=ielp0,tv=elp)
  close(iunit)

  cbuf="data2.out"
  open(iunit,FILE=cbuf) 
  elp0 = get_cpu_time()  
  do k = 1, NZ
  do j = 1, NY
  do i = 1, NX
     write(iunit,'(1I10)') AA(i,j,k)
  end do
  end do
  end do
  cpu = get_cpu_time() - elp0 
  close(iunit)

  !-------------------------------------------------------------------
  ! write with unformat
  !-------------------------------------------------------------------
  else if ( ifmt .eq. 1 ) then

  cbuf="data1.out"
  open(iunit,FILE=cbuf,FORM='unformatted',ACCESS='stream') 
  call get_elp_time(ioffset=ielp0)  
  write(iunit) AA
  call get_elp_time(ioffset=ielp0,tv=elp)
  close(iunit)
 
  cbuf="data2.out"
  open(iunit,FILE=cbuf,FORM='unformatted',ACCESS='stream') 
  elp0 = get_cpu_time()  
  write(iunit) AA
  cpu = get_cpu_time() - elp0
  close(iunit)

  !-------------------------------------------------------------------
  ! others 
  !-------------------------------------------------------------------
  else
 
  write(6,'(5x,1A)') "[Error] No routine. Check command-line options."
  deallocate( AA )
  stop

  end if

  write(6,'(1A)') "format memory_MiB elapsed_time_sec cpu_time_sec diff"
  write(6,'(1I3,1F11.4,1F11.4,1F11.4,1F11.4)') &
&   ifmt, aamemsize, elp, cpu, elp-cpu

  deallocate( AA )

  stop
end program

File: Tuning/sample_code/15_io-format/src/cpp/main.cpp
// Copyright 2024 Research Organization for Information Science and Technology
#include <iostream>
#include <fstream>
#include <ios>
#include <iomanip>
#include <cstdlib>
#include "memory.h"
#include "timer.h"

int main(int argc, char **argv)
{
  const int nx=1024; // 2^10
  const int ny=1024; // 2^10
  int nz;
  int i, j, k;
  int ifmt;
  int aasize;

  double elp, cpu, elp0;
  double aamemsize;

  if ( argc != 3 ) {
     std::cout << "[usage] ./run.x (arg1) (arg2)                 \n"
               << "   (arg1): Integer (0-1)                      \n"
	       << "           0: write w/ format                 \n"
	       << "           1: write w/o format                \n"
	       << "   (arg2): Integer (<200) to change array size\n";
     return EXIT_SUCCESS;
  }

  ifmt = atoi(argv[1]);
  nz = atoi(argv[2]);

  if ( nz > 200 ) {
     std::cout << "[Error] 2nd arg must be < 200." << std::endl;
     return EXIT_FAILURE;
  }

  int ***aa;
  create_3d_array( aa, nx, ny, nz);

  for ( i=0; i<nx; ++i ) {
  for ( j=0; j<ny; ++j ) {
  for ( k=0; k<nz; ++k ) {
      aa[i][j][k] = k + (j + i*ny)*nz;
  }}}

  aasize = nx*ny*nz;
  aamemsize = 4.0*nx*ny*nz / 1048576; // MiB

  //-----------------------------------------------------------------
  //  write with format
  //-----------------------------------------------------------------
  if ( ifmt == 0 ) {

    std::ofstream ofs1("data1.out");
    elp0 = get_elp_time();
    for ( i=0; i<nx; ++i ) {
    for ( j=0; j<ny; ++j ) {
    for ( k=0; k<nz; ++k ) {
       ofs1 << std::dec 
            << aa[i][j][k] << std::endl;
    }}}
    elp = get_elp_time() - elp0;
    ofs1.close();

    std::ofstream ofs2("data2.out");
    elp0 = get_cpu_time();
    for ( i=0; i<nx; ++i ) {
    for ( j=0; j<ny; ++j ) {
    for ( k=0; k<nz; ++k ) {
       ofs2 << std::dec 
            << aa[i][j][k] << std::endl;
    }}}
    cpu = get_cpu_time() - elp0;
    ofs2.close();

  //-----------------------------------------------------------------
  // write without format
  //-----------------------------------------------------------------
  } else if ( ifmt == 1 ) {

    std::ofstream ofs1("data1.out",std::ios_base::binary);
    elp0 = get_elp_time();
    ofs1.write(reinterpret_cast<const char*>(&aa[0][0][0]), aasize*sizeof(int));
    elp = get_elp_time() - elp0;
    ofs1.close();

    std::ofstream ofs2("data2.out",std::ios_base::binary);
    elp0 = get_cpu_time();
    ofs2.write(reinterpret_cast<const char*>(&aa[0][0][0]), aasize*sizeof(int));
    cpu = get_cpu_time() - elp0;
    ofs2.close();

  } else {
    std::cout <<"[Error] No routine. Check command-line options" << std::endl;
    release_3d_array(aa);
    return EXIT_FAILURE;
  }

  std::cout << "format memory_MiB elapsed_sec cpu_time_sec diff\n"
            << std::dec 
            << std::fixed
	    << std::setw(3)
	    << ifmt << " "
	    << std::setw(10)
	    << std::setprecision(3) 
	    << aamemsize << " "
	    << elp << " "
	    << cpu << " " 
	    << elp - cpu << "\n";

  release_3d_array(aa);

  return EXIT_SUCCESS;
}
